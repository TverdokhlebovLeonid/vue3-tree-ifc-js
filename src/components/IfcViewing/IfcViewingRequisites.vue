<script setup lang="ts">
import { ref } from 'vue'
import type { IModelLevels, IDataLevelHide } from '@/components/IfcViewing/interfaceIfcViewing'
import { dictionaryIfcViewing, DEFAULT_KEY } from '@/components/IfcViewing/dictionaryIfcViewing'

const modelLevels = ref<IModelLevels[]>([])
const levelActive = ref<number | null>(null)

const emits = defineEmits<{
  (e: 'transfer-level', value: number[]): void
  (e: 'level-hide', value: IDataLevelHide): void
}>()

const transferLevel = (children: IModelLevels[], id: number): void => {
  const level: number[] = []
  if (levelActive.value === id) {
    levelActive.value = null
  } else {
    levelActive.value = id
    children.length
      ? children.forEach((el: IModelLevels) => {
          level.push(el.expressID)
        })
      : level.push(id)
  }
  emits('transfer-level', level)
}

const setLevelHide = (customID: string, check: boolean): void => {
  emits('level-hide', { customID, check })
}
const setModel = (model: IModelLevels[]): void => {
  levelActive.value = null
  modelLevels.value = model
}

const setElementName = (name: string): string => {
  const key = dictionaryIfcViewing?.[name] ? name : DEFAULT_KEY
  return dictionaryIfcViewing[key].single
}

defineExpose({
  setModel,
  modelLevels,
  setElementName,
  levelActive,
  transferLevel,
  setLevelHide,
})

const PROPS_TREE = {
  value: 'expressID',
  label: 'type',
  children: 'children',
}
</script>

<template>
  <div v-if="modelLevels.length" class="ifc-requisites">
    <el-text size="large" tag="b" type="primary">Реквизиты</el-text>
    <el-tree-v2
      style="max-width: 100%"
      :expand-on-click-node="false"
      :data="modelLevels"
      :props="PROPS_TREE"
      :height="400"
    >
      <template #default="{ data }">
        <el-tooltip
          v-if="data?.customID"
          effect="dark"
          :content="data.check ? 'Удалить элемент' : 'Вернуть элемент'"
          placement="bottom"
        >
          <el-checkbox
            v-model="data.check"
            size="large"
            @change="setLevelHide(data.customID, data.check)"
          />
        </el-tooltip>
        <el-tooltip
          effect="dark"
          :content="data.expressID === levelActive ? 'Убрать выделение' : 'Выделить элемент'"
          placement="bottom"
        >
          <el-text
            :type="data.expressID === levelActive ? 'danger' : 'primary'"
            @click="transferLevel(data?.children || [], data.expressID)"
          >
            {{ setElementName(data?.type?.toString() || DEFAULT_KEY) }}
          </el-text>
        </el-tooltip>
      </template>
    </el-tree-v2>
  </div>
</template>

<style scoped lang="sass">
.ifc-requisites
  margin-top: 20px
  .el-tree
    background-color: $color-bg-light
    margin-top: 10px
</style>
