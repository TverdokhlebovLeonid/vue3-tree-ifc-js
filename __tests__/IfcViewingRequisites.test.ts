import { mount } from '@vue/test-utils'
import { expect, test } from 'vitest'
import IfcViewingRequisites from '@/components/IfcViewing/IfcViewingRequisites.vue'
import { nextTick } from 'vue'
import { modelLevels } from './dataMock'

const TEST_ID = 63

test('Snapshot and render component', async () => {
  const component = mount(IfcViewingRequisites)
  component.vm.modelLevels = modelLevels
  await nextTick()
  expect(component.find('.ifc-requisites').find('b').text()).toBe('Реквизиты')
  expect(component.html()).toMatchSnapshot()
})

test('Function setElementName', async () => {
  const component = mount(IfcViewingRequisites)
  const func = component.vm.setElementName
  const name = 'IFCBUILDINGSTOREY'
  expect(func(name)).toBe('Этаж/Уровень')
  const nameDefault = 'NO_SUCH_NAME'
  expect(func(nameDefault)).toBe('Элемент')
})

test('Function setModel', async () => {
  const component = mount(IfcViewingRequisites)
  component.vm.setModel(modelLevels)
  expect(component.vm.modelLevels[0].customID).toBe('0-level')
  expect(component.vm.levelActive).toBe(null)
})

test('Function transferLevel', async () => {
  const component = mount(IfcViewingRequisites)
  component.vm.levelActive = TEST_ID
  component.vm.transferLevel(modelLevels, TEST_ID)
  expect(component.vm.levelActive).toBe(null)
  expect(component.emitted('transfer-level')?.[0]).toEqual([[]])
})

test('Function transferLevel(same id)', async () => {
  const component = mount(IfcViewingRequisites)
  component.vm.levelActive = 50
  component.vm.transferLevel(modelLevels, TEST_ID)
  expect(component.vm.levelActive).toBe(TEST_ID)
  expect(component.emitted('transfer-level')?.[0]).toEqual([[TEST_ID]])
})

test('Function setLevelHide', async () => {
  const component = mount(IfcViewingRequisites)
  component.vm.setLevelHide('50', true)
  expect(component.emitted('level-hide')?.[0]).toEqual([{ customID: '50', check: true }])
})
