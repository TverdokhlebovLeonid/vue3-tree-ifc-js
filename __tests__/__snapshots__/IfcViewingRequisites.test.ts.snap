// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IfcViewingRequisites test component > Snapshot and render component 1`] = `
"<div data-v-3063711b="" class="ifc-requisites"><b data-v-3063711b="" class="el-text el-text--primary el-text--large">Реквизиты</b>
  <div data-v-3063711b="" class="el-tree" role="tree" style="max-width: 100%;">
    <div class="el-tree__empty-block"><span class="el-tree__empty-text">No Data</span></div>
  </div>
</div>"
`;
